File: frontend/src/lib/seriesEvidence.ts
// Pure helpers of the Series lens's LANES stage (SERIES ARC S5): the strip's
// metrics as overlay-chart series (one per visible lane, frame-index x, a
// lane's missing fits break its line), the playhead drawn on that chart (a
// vertical two-point series + one marker per lane at the playhead's value),
// the evidence table's columns with the "best lane" rule, and the filter
// ring's cursor step for the playhead frame. No React — vitest-locked.
import type { OverlayMarker, OverlaySeries } from "../components/OverlayCurvesChart";
import { formatPct } from "./chartScale";
import { laneLabel, laneStyle } from "./seriesLanes";
import type { LaneEvidence, LaneSpec, StripMetric, StripPayload } from "./seriesTypes";

/** A strip metric or the ATM vol (carried beside the metrics on the strip). */
export type EvidenceMetric = StripMetric | "atmVol";

/** The metric chip order of the Lanes stage. */
export const EVIDENCE_METRICS: readonly EvidenceMetric[] = [
  "rmsBp", "maxIvBp", "pullAtmBp", "zetaAtm", "fitMs", "atmVol", "skew", "curvature",
];

const METRIC_LABELS: Record<EvidenceMetric, string> = {
  rmsBp: "rms bp",
  maxIvBp: "max err bp",
  pullAtmBp: "pull ATM bp",
  zetaAtm: "ζ ATM",
  fitMs: "fit ms",
  atmVol: "ATM σ",
  skew: "skew",
  curvature: "curvature",
};

const METRIC_DIGITS: Record<EvidenceMetric, number> = {
  rmsBp: 1, maxIvBp: 1, pullAtmBp: 1, zetaAtm: 2, fitMs: 0, atmVol: 1, skew: 3, curvature: 3,
};

/** The filmstrip's playhead violet (components/series/Filmstrip). */
export const PLAYHEAD_COLOUR = "rgb(167 139 250 / 0.95)";
export const PLAYHEAD_LABEL = "playhead";

/** The chip / axis label of a metric. */
export function metricLabel(metric: EvidenceMetric): string {
  return METRIC_LABELS[metric];
}

/** A metric value for a readout: the ATM vol as a percentage, the rest at
 *  the metric's own precision; "—" when null / non-finite. */
export function formatMetric(metric: EvidenceMetric, v: number | null | undefined): string {
  if (v == null || !Number.isFinite(v)) return "—";
  return metric === "atmVol" ? formatPct(v, METRIC_DIGITS.atmVol) : v.toFixed(METRIC_DIGITS[metric]);
}

/** One lane's per-frame values of a metric off the strip (empty when absent). */
export function metricValues(strip: StripPayload, metric: EvidenceMetric, laneId: string): (number | null)[] {
  return (metric === "atmVol" ? strip.atmVol[laneId] : strip.lanes[laneId]?.[metric]) ?? [];
}

export interface LaneOrdinal {
  lane: LaneSpec;
  ordinal: number;
}

type StyleOf = (lane: LaneSpec, ordinal: number) => { colour: string; dash: string };

/** One overlay series per lane: x = the frame's POSITION in the strip (the
 *  playback index), y = the metric (NaN where the lane has no fit, which
 *  breaks the line). The lane's colour + ordinal dash; a duplicated label
 *  carries the lane id (the chart keys its paths by label). */
export function stripSeries(
  strip: StripPayload,
  metric: EvidenceMetric,
  lanes: readonly LaneOrdinal[],
  style: StyleOf = laneStyle,
): OverlaySeries[] {
  const n = strip.idx.length;
  const labels = lanes.map(({ lane }) => laneLabel(lane));
  return lanes.map(({ lane, ordinal }, j) => {
    const values = metricValues(strip, metric, lane.id);
    const st = style(lane, ordinal);
    const duplicate = labels.some((l, k) => k !== j && l === labels[j]);
    const xs: number[] = [];
    const ys: number[] = [];
    for (let i = 0; i < n; i++) {
      const v = values[i];
      xs.push(i);
      ys.push(v != null && Number.isFinite(v) ? v : Number.NaN);
    }
    return {
      label: duplicate ? `${labels[j]} (${lane.id})` : labels[j],
      xs,
      ys,
      color: st.colour,
      dash: st.dash !== "" ? st.dash : undefined,
    };
  });
}

/** The playhead as a vertical two-point series spanning the lanes' finite
 *  y-extent at `index` (null when nothing is finite — no line to anchor). */
export function playheadLine(index: number, series: readonly OverlaySeries[]): OverlaySeries | null {
  let lo = Infinity;
  let hi = -Infinity;
  for (const s of series) {
    for (const y of s.ys) {
      if (!Number.isFinite(y)) continue;
      if (y < lo) lo = y;
      if (y > hi) hi = y;
    }
  }
  if (!(lo <= hi)) return null;
  if (lo === hi) {
    const half = Math.max(Math.abs(lo) * 0.05, 1e-6);
    lo -= half;
    hi += half;
  }
  return { label: PLAYHEAD_LABEL, xs: [index, index], ys: [lo, hi], color: PLAYHEAD_COLOUR, dash: "3 3" };
}

/** One marker per lane series at the playhead: the lane's value at frame
 *  `index` (skipped where the lane has no fit there), titled "lane · value". */
export function playheadMarker(
  index: number,
  series: readonly OverlaySeries[],
  fmt: (v: number) => string = (v) => v.toFixed(2),
): OverlayMarker[] {
  const out: OverlayMarker[] = [];
  for (const s of series) {
    if (s.label === PLAYHEAD_LABEL) continue;
    const i = s.xs.indexOf(index);
    if (i < 0) continue;
    const y = s.ys[i];
    if (!Number.isFinite(y)) continue;
    out.push({ x: index, y, label: `${s.label} · ${fmt(y)}`, color: s.color });
  }
  return out;
}

/** The ring step the playhead sits on, matched by FRAME INDEX — the lane
 *  filter payload's `frameIdx[i]` is the series frame the i-th step was
 *  committed on (a step's `ts` is a local-clock epoch of a naive stamp, so
 *  steps are never matched to frames by clock): the step whose frameIdx is
 *  the playhead index, else the LAST step committed at or before it, else
 *  null (the ring starts after the frame). */
export function cursorStepIndex(frameIdx: readonly (number | null)[], index: number): number | null {
  let before: number | null = null;
  for (let i = 0; i < frameIdx.length; i++) {
    const f = frameIdx[i];
    if (f == null || !Number.isFinite(f)) continue;
    if (f === index) return i;
    if (f < index) before = i;
  }
  return before;
}

/** Fixed-point evidence number; "—" for null / non-finite. */
export function fmtEvidence(v: number | null | undefined, digits: number): string {
  return v == null || !Number.isFinite(v) ? "—" : v.toFixed(digits);
}

/** A numeric column of the evidence table. `score` maps a value to "lower is
 *  better" (identity by default — every rms / roughness / cost column). */
export interface EvidenceColumn {
  key: Exclude<keyof LaneEvidence, "nFrames" | "nFailed" | "worstFrame">;
  label: string;
  title: string;
  digits: number;
  score?: (v: number) => number;
}

export const EVIDENCE_COLUMNS: readonly EvidenceColumn[] = [
  { key: "meanRmsBp", label: "mean rms bp", title: "mean rms of the fit to the quotes over the ready frames (vol bp)", digits: 1 },
  { key: "meanMaxIvBp", label: "mean max bp", title: "mean of the worst quote error per frame (vol bp)", digits: 1 },
  { key: "roughnessAtmBp", label: "rough. ATM bp", title: "mean |Δ σ_atm| between consecutive frames (vol bp)", digits: 1 },
  { key: "roughnessSkew", label: "rough. skew", title: "mean |Δ skew| between consecutive frames", digits: 4 },
  { key: "meanAbsPullAtmBp", label: "mean |pull| bp", title: "mean |ATM pull| of the prior / filter (vol bp)", digits: 1 },
  { key: "zetaAtmStd", label: "ζ ATM std", title: "std of the ATM standardized innovation — 1 is a healthy Q", digits: 2, score: (v) => Math.abs(v - 1) },
  { key: "meanFitMs", label: "mean fit ms", title: "mean wall time of one frame's fit", digits: 0 },
];

/** The lane with the best value of a column among `laneIds` — null unless at
 *  least two lanes carry a finite value (one row has nothing to beat). */
export function bestLane(
  evidence: Record<string, LaneEvidence>,
  laneIds: readonly string[],
  col: EvidenceColumn,
): string | null {
  const score = col.score ?? ((v: number) => v);
  let best: string | null = null;
  let bestScore = Infinity;
  let finite = 0;
  for (const id of laneIds) {
    const v = evidence[id]?.[col.key];
    if (v == null || !Number.isFinite(v)) continue;
    finite += 1;
    const s = score(v);
    if (s < bestScore) {
      bestScore = s;
      best = id;
    }
  }
  return finite >= 2 ? best : null;
}

File: backend/volfit/api/series_evidence.py
"""The Lanes stage's evidence (SERIES ARC S5, roadmap §3.4 "Lanes").

``evidence_payload`` summarizes every lane over the READY frames of the shown
expiry, derived from the same strip the filmstrip draws (one source, the
numbers agree): the mean rms / max error in vol bp, the worst frame, the
HANDLE-PATH ROUGHNESS — the mean frame-to-frame move of the ATM vol (bp)
and of the skew, i.e. what a prior or a filter damps, read beside the rms
it costs —, the mean (and mean absolute) pull against the free reference,
the fit time and the spread of the filter's ATM ζ. ``lane_filter_index`` /
``lane_filter_ring`` serve the lane's observation-filter rings from its
carry document (``series_lanes.filter_json``: the ``filterHistory`` entries
in the workspace-doc shape, each step already the ``/filter/history`` wire
dict), so the FilterTimeline draws a lane's ring as it draws the live one.
"""

from __future__ import annotations

import numpy as np
from pydantic import BaseModel

from volfit.api.series_payload import UnknownSeriesError, strip_payload
from volfit.api.series_store import SeriesStore
from volfit.data.store import VolStore


class WorstFrame(BaseModel):
    idx: int
    ts: str
    rmsBp: float


class LaneEvidence(BaseModel):
    nFrames: int = 0
    nFailed: int = 0
    meanRmsBp: float | None = None
    meanMaxIvBp: float | None = None
    worstFrame: WorstFrame | None = None
    roughnessAtmBp: float | None = None
    roughnessSkew: float | None = None
    meanPullAtmBp: float | None = None
    meanAbsPullAtmBp: float | None = None
    meanFitMs: float | None = None
    zetaAtmStd: float | None = None


class EvidencePayload(BaseModel):
    seriesId: str
    expiry: str | None = None
    lanes: dict[str, LaneEvidence] = {}


class LaneFilterIndex(BaseModel):
    laneId: str
    expiries: list[str] = []


class LaneFilterPayload(BaseModel):
    laneId: str
    expiry: str
    steps: list[dict] = []
    #: The series frame each step was committed on, by ORDER (the ring keeps
    #: the last 64 steps; a step's ``ts`` is the app's local-clock epoch of a
    #: naive timestamp, so the frame is matched by order, never by clock).
    frameIdx: list[int | None] = []


# --------------------------------------------------------------- evidence

def _mean(values) -> float | None:
    xs = [float(v) for v in values if v is not None and np.isfinite(v)]
    return round(float(np.mean(xs)), 4) if xs else None


def _roughness(values, scale: float = 1.0) -> float | None:
    """Mean |Δ| between CONSECUTIVE frames that both carry a value."""
    diffs = [abs(float(b) - float(a)) * scale for a, b in zip(values, values[1:])
             if a is not None and b is not None]
    return round(float(np.mean(diffs)), 4) if diffs else None


def _std(values) -> float | None:
    xs = [float(v) for v in values if v is not None and np.isfinite(v)]
    return round(float(np.std(xs)), 4) if len(xs) > 1 else None


def evidence_payload(state, series_id: str, lane_ids: list[str] | None = None,
                     expiry: str | None = None) -> EvidencePayload:
    """GET /series/{id}/evidence?lanes=&expiry= — per lane, over the strip."""
    strip = strip_payload(state, series_id, lane_ids, expiry)
    with VolStore(state.store_path) as store:
        series = SeriesStore(store)
        doc = series.get(series_id)
        if doc is None:
            raise UnknownSeriesError(series_id)
        failed = {lane_id: {f.idx for f in series.fits(series_id, lane_id=lane_id)
                            if f.status == "failed"} for lane_id in strip.lanes}
    shown = expiry or (next((f.expiries[0] for f in doc.frames if f.expiries), None))
    lanes: dict[str, LaneEvidence] = {}
    for lane_id, metrics in strip.lanes.items():
        rms = metrics["rmsBp"]
        worst = None
        scored = [(v, i) for i, v in enumerate(rms) if v is not None]
        if scored:
            v, i = max(scored)
            worst = WorstFrame(idx=strip.idx[i], ts=strip.ts[i], rmsBp=float(v))
        lanes[lane_id] = LaneEvidence(
            nFrames=sum(1 for v in rms if v is not None),
            nFailed=len(failed.get(lane_id, ())),
            meanRmsBp=_mean(rms), meanMaxIvBp=_mean(metrics["maxIvBp"]), worstFrame=worst,
            roughnessAtmBp=_roughness(strip.atmVol.get(lane_id, []), 1e4),
            roughnessSkew=_roughness(metrics["skew"]),
            meanPullAtmBp=_mean(metrics["pullAtmBp"]),
            meanAbsPullAtmBp=_mean([abs(v) for v in metrics["pullAtmBp"] if v is not None]),
            meanFitMs=_mean(metrics["fitMs"]), zetaAtmStd=_std(metrics["zetaAtm"]),
        )
    return EvidencePayload(seriesId=series_id, expiry=shown, lanes=lanes)


# ------------------------------------------------------------ filter rings

def _rings(state, series_id: str, lane_id: str) -> list[dict]:
    with VolStore(state.store_path) as store:
        series = SeriesStore(store)
        if not series.exists(series_id):
            raise UnknownSeriesError(series_id)
        if lane_id not in {lane.id for lane in series.lanes(series_id)}:
            raise KeyError(lane_id)
        carry = series.lane_filter(series_id, lane_id) or {}
    return list(carry.get("filterHistory") or [])


def lane_filter_index(state, series_id: str, lane_id: str) -> LaneFilterIndex:
    rings = _rings(state, series_id, lane_id)
    return LaneFilterIndex(laneId=lane_id, expiries=sorted({r["expiry"] for r in rings
                                                            if r.get("steps")}))


def lane_filter_ring(state, series_id: str, lane_id: str, expiry: str) -> LaneFilterPayload:
    """The lane's ring for one node (any fit mode; the lane runs one)."""
    rings = _rings(state, series_id, lane_id)
    steps = list(next((r.get("steps") or [] for r in rings if r.get("expiry") == expiry), []))
    with VolStore(state.store_path) as store:
        done = sorted({f.idx for f in SeriesStore(store).fits(series_id, lane_id=lane_id)
                       if f.expiry == expiry and f.status == "done"})
    frames: list[int | None] = list(done[-len(steps):]) if steps else []
    if len(frames) < len(steps):  # more steps than stored frames (never expected): pad left
        frames = [None] * (len(steps) - len(frames)) + frames
    return LaneFilterPayload(laneId=lane_id, expiry=expiry, steps=steps, frameIdx=frames)
